mirgecom/initializers.py

```diff
@@ -462,6 +450,218 @@ def exact_rhs(self, discr, q, t=0.0):
+                   len(spec_centers) != nspecies:
+                raise ValueError(f"Expected nspecies={nspecies} inputs.")
+            for i in range(nspecies):
+                if len(spec_centers[i]) != dim:
+                    raise ValueError(f"Expected {dim}-dimensional "
+                                     f"inputs for spec_centers.")
+
+        self._nspecies = nspecies
+        self._dim = dim
+        self._velocity = velocity
+        self._center = center
+        self._p0 = p0
+        self._rho0 = rho0
+        self._spec_y0s = spec_y0s
+        self._spec_centers = spec_centers
+        self._spec_amplitudes = spec_amplitudes
+
+    def __call__(self, x_vec, *, t=0, eos=IdealSingleGas()):
+        """
+        Create a multi-component lump solution at time *t* and locations *x_vec*.
+
+        Note that *t* is used to advect the component lumps under the assumption of
+        constant, and uniform velocity.
+
+        Parameters
+        ----------
+        t: float
+            Current time at which the solution is desired
+        x_vec: numpy.ndarray
+            Nodal coordinates
+        eos: :class:`mirgecom.eos.GasEOS`
+            Equation of state class to be used in construction of soln (if needed)
+        """
+        if len(x_vec) != self._dim:
+            print(f"len(x_vec) = {len(x_vec)}")
+            print(f"self._dim = {self._dim}")
+            raise ValueError(f"Expected {self._dim}-dimensional inputs.")
+
+        actx = x_vec[0].array_context
+
+        # mass0 = self._rho0
+        amplitude = 0.0
+        loc_update = t * self._velocity
+        lump_loc = self._center + loc_update
+        # coordinates relative to lump center
+        rel_center = make_obj_array(
+            [x_vec[i] - lump_loc[i] for i in range(self._dim)]
+        )
+        r = actx.np.sqrt(np.dot(rel_center, rel_center))
+        expterm = amplitude * actx.np.exp(- r ** 2)
+        mass = expterm + self._rho0
+        mom = self._velocity * make_obj_array([mass])
+        gamma = eos.gamma()
+        energy = (self._p0 / (gamma - 1.0)) + np.dot(mom, mom) / (2.0 * mass)
+
+        # process the species components independently
+        mass_fracs = None
+        if self._nspecies > 0:
+            lump_locs = make_obj_array([self._spec_centers[i] + loc_update
+                                       for i in range(self._nspecies)])
+            rel_poss = make_obj_array([x_vec - lump_locs[i]
+                                       for i in range(self._nspecies)])
+            r2s = make_obj_array([np.dot(rel_poss[i], rel_poss[i])
+                                  for i in range(self._nspecies)])
+            expterms = make_obj_array(
+                [self._spec_amplitudes[i] * actx.np.exp(- r2s[i])
+                 for i in range(self._nspecies)]
+            )
+            mass_fracs = make_obj_array([self._spec_y0s[i] + expterms[i]
+                                        for i in range(self._nspecies)])
```

Comment:
Great question.   The use case is testing that each scalar component does the right thing separately.   The purpose of the Multicomponent lump initializer is to add multiple scalar lumps in disparate locations for testing purposes.  Having them in separate locations avoids any potential for getting the final solution (i.e. after advecting the lumps around) right by fluke.

Reply:
Different amplitudes should be sufficient for that, I would think?

Reply:
:thinking:
Since the RHS is a function of r_alpha, I wanted to makes sure that I had a distinct r_alpha for each component.  I do see that using different amplitudes for each component would be sufficient for making the RHS vary across components.